pkg/builder/service_factory.go

```diff
@@ -0,0 +1,172 @@
+package builder
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+
+	"github.com/aserto-dev/certs"
+	metrics "github.com/aserto-dev/go-http-metrics/middleware/grpc"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/pkg/errors"
+	"github.com/rs/cors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type ServiceFactory struct{}
+
+func NewServiceFactory() *ServiceFactory {
+	return &ServiceFactory{}
+}
+
+func (f *ServiceFactory) CreateService(config *API, opts []grpc.ServerOption, registrations GRPCRegistrations,
+	handlerRegistrations HandlerRegistrations,
+	withGateway bool) (*Server, error) {
+	grpcServer, err := prepareGrpcServer(&config.GRPC.Certs, opts)
+	if err != nil {
+		return nil, err
+	}
+	registrations(grpcServer)
+	reflection.Register(grpcServer)
+
+	listener, err := net.Listen("tcp", config.GRPC.ListenAddress)
+	if err != nil {
+		return nil, err
+	}
+	gate := Gateway{}
+	if withGateway && config.Gateway.ListenAddress != "" {
+		gate, err = prepareGateway(config, handlerRegistrations)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var health *HealthServer
+	if config.Health.ListenAddress != "" {
+		health = newGRPCHealthServer()
+	}
+
+	return &Server{
+		Config:        config,
+		Server:        grpcServer,
+		Listener:      listener,
+		Registrations: registrations,
+		Gateway:       gate,
+		Health:        health,
+		Started:       make(chan bool),
+	}, nil
+}
+
+func prepareGateway(config *API, registrations HandlerRegistrations) (Gateway, error) {
+	c := cors.New(cors.Options{
+		AllowedHeaders: []string{"Authorization", "Content-Type", "Depth"},
+		AllowedOrigins: config.Gateway.AllowedOrigins,
+		AllowedMethods: []string{http.MethodGet, http.MethodPost, http.MethodHead, http.MethodDelete, http.MethodPut,
+			http.MethodPatch, "PROPFIND", "MKCOL", "COPY", "MOVE"},
+		Debug: true,
+	})
+
+	runtimeMux := gatewayMux()
+
+	tlsCreds, err := certs.GatewayAsClientTLSCreds(config.GRPC.Certs)
+	if err != nil {
+		return Gateway{}, errors.Wrapf(err, "failed to get TLS credentials")
+	}
+	grpcEndpoint := fmt.Sprintf("dns:///%s", config.GRPC.ListenAddress)
+
+	opts := []grpc.DialOption{
+		grpc.WithTransportCredentials(tlsCreds),
+	}
+
+	err = registrations(context.Background(), runtimeMux, grpcEndpoint, opts)
+	if err != nil {
+		return Gateway{}, err
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle("/", runtimeMux)
+	mux.Handle("/api/", fieldsMaskHandler(runtimeMux))
```

Comment:
Maybe not an immediate blocker, but we should provide a way to expose OpenAPI specs services that have a REST API.

Reply:
For now the service's gateway mux is public after the service creation and I'm currently attaching the openapi handlers in the authorizer initialization, but I think we can improve on this.